packages/labs/router/src/router.ts

```diff
@@ -4,4 +4,82 @@
  * SPDX-License-Identifier: BSD-3-Clause
  */
 
-export {};
+import {ReactiveControllerHost} from 'lit';
+import {Routes, RouteConfig} from './routes.js';
+
+/**
+ * A root-level router that installs global event listeners to intercept
+ * navigation.
+ *
+ * This class extends Routes so that it can also have a route configuration.
+ *
+ * There should only be one Router instance on a page, since the Router
+ * installs global event listeners on `window` and `document`. Nested
+ * routes should be configured with the `Routes` class.
+ */
+export class Router extends Routes {
+  constructor(
+    host: ReactiveControllerHost & HTMLElement,
+    routes?: Array<RouteConfig>
+  ) {
+    super(host, routes ?? []);
+    // Kick off routed rendering by going to the current URL
+    // TODO: should this be done in the first update instead?
+    this.goto(window.location.pathname);
+  }
+
+  override hostConnected() {
+    super.hostConnected();
+    window.addEventListener('click', this._onClick);
+    window.addEventListener('popstate', this._onPopState);
+  }
+
+  override hostDisconnected() {
+    super.hostDisconnected();
+    window.removeEventListener('click', this._onClick);
+    window.removeEventListener('popstate', this._onPopState);
+  }
+
+  private _onClick = (e: MouseEvent) => {
+    const isNonNavigationClick =
```

Comment:
Crazy that all this low-level stuff has to get checked! Feels like there are some missing platform APIs for intercepting navigations.

Reply:
There's a really nice App History API in origin trial right now!